### DISCUSSION TITLE
Add "Back" link to Invalid SQL page

### DISCUSSION THREAD

**Discussion by elgiad007** [simonw/datasette#1334] [Ideas] [answered]
I very often find myself falling down the rabbit hole with custom SQL queries on a new and interesting table. When I make a mistake in my custom SQL query in Datasette (which is often, especially when accessing Datasette on my phone), I often find that the "Back" button on my browser doesn't always restore the SQL code I was working on, resulting in lost work. I'm not sure what circumstances lead to the bad SQL being restored or not, but it does not seem consistent across browsers (mostly been using Chrome, but on different versions of Windows and Android).

It would be nice to have a link on the "Invalid SQL" page that brought me back to the previous page but with my SQL code (however erroneous) restored.  The bad SQL is still in the address bar, so I know it's still available, I just can't rely on my web browser to restore it correctly.

**Comment by brandonrobertz** [comment]
Right now the SQL in that box is just handled like any normal textarea box by your browser. Some browsers, in some situations, can hold onto the old state of the textarea box and sometimes it won't. This isn't a a Datasette problem, per-se, it's just browser behavior.

A work around would be to use a plugin like [datasette-saved-queries](https://datasette.io/plugins/datasette-saved-queries) or write one yourself that listens for input on that box and saves it someway (LocalStorage, etc).

**Reply by elgiad007** [reply]
Hi Brandon,

I apologize if my submission wasn't in the appropriate place (thought I had submitted it to an 'ideas' page).  I wasn't trying to suggest that this was a problem with Datasette, but rather something to enhance usability for end users who aren't in charge of their own Datasette instances and are working with someone else's installation that may not have the saved-queries plugin. 

The saved-queries plugin appears to be useful for submitting finished queries that we don't intend to change again, preferably ones that are known to work and do not have syntax or logic errors.  The enhancement I'm suggesting is more for when you're still developing the query in the browser. Since browsers are known to be unreliable when it comes to re-populating a text box when the user clicks the Back button, I thought adding a link to do this on the the "Invalid SQL" page might do the trick. It seemed like a fairly small change that could have a big impact on usability.

I hope that clarifies things. I'm finding more and more ways to use Datasette both personally and professionally and am very excited to see the project grow

Luc.

**Reply by brandonrobertz** [reply]
Oh yeah, I agree. A plugin like this would be super handy and not that hard to build. I've also found myself frustrated by having my query disappear into the void. Now I write my queries in an actual code editor and copy/paste them into the text box as a result.

**Comment by simonw** [comment]
This is a long-standing horrible user experience bug - I've had an issue open since November 2019 for it! https://github.com/simonw/datasette/issues/619

It's weirdly difficult to fix (I really need to refactor how that part of the code works entirely) but it's also a big pain point, so it may be worth me diving in and figuring out an unpleasant fix rather than waiting for the refactor.

**Reply by elgiad007** [reply]
Thanks Simon! Keep up the great work. This is an extremely useful tool.

**Reply by simonw** [reply]
I fixed it! Here's a demo: https://latest.datasette.io/fixtures?sql=select%0D%0A++*%0D%0Afrom%0D%0A++[foo]

See https://github.com/simonw/datasette/issues/619 for details.

**Reply by brandonrobertz** [reply]
Oh Snap this is such a huge UX improvement!!

**Comment by elgiad007** [comment]
Beautiful! Thank you very much for this. I can now make endless syntax
errors without worrying about losing my work 👍

Thanks again,
Luc

**Comment by simonw** [comment, marked as answer]
This is now released in Datasette 0.57!

**Comment by elgiad007** [comment]
Great, thanks!

Luc
